Database: Add autoincrement, uniqueness, and sync-writes polyfills (#28)
crate/sqlalchemy-cratedb: docs/support.md

(support-features)=
# Support Features

The package bundles a few support and utility functions that try to fill a few
gaps you will observe when working with CrateDB, a distributed OLAP database,
because it lacks certain features traditional OLTP databases include.

> The features outlined below are referred to as poly-fills, and emulate a few

Author: Apparently it's one word. Thanks.

-- https://en.wikipedia.org/wiki/Polyfill_(programming)

will simply assign `sa.func.now()` as a column `default`. You can use it if
adjusting ORM models is out of reach for your database adapter at hand.

The patch enables to optionally use `autoincrement=True` on column definitions.
It works on SQLAlchemy column types `sa.BigInteger`, `sa.DateTime`, and
`sa.String`.

```python
import sqlalchemy as sa
from sqlalchemy.orm import declarative_base
from sqlalchemy_cratedb.support import patch_autoincrement_timestamp

# Enable patch.
patch_autoincrement_timestamp()

# Define database schema.
Base = declarative_base()

class FooBar(Base):
    id = sa.Column(sa.DateTime, primary_key=True, autoincrement=True)
```

## Synthetic UNIQUE Constraints

CrateDB does not provide `UNIQUE` constraints. Because of its distributed
nature, it would be an expensive operation. This feature emulates corresponding
functionality by querying the table for unique values before invoking the SQL
`INSERT` operation. 

When the uniqueness constraint is violated, the adapter will raise a
corresponding exception.
```python
IntegrityError: DuplicateKeyException in table 'foobar' on constraint 'name'
```

> Please note querying the table before each insert operation can cause serious

Author: Good idea, thanks.

There are situations where stronger consistency is required, for example when
needing to satisfy test suites of 3rd party frameworks, which usually do not
take such special behavior of CrateDB into consideration. 

:::{rubric} Utility
:::
The `refresh_after_dml` utility will configure an SQLAlchemy engine or session
to automatically invoke relevant `REFRESH TABLE` statements after each DML
operation (INSERT, UPDATE, DELETE), for the corresponding entities / tables.

```python
import sqlalchemy as sa
from sqlalchemy_cratedb.support import refresh_after_dml

engine = sa.create_engine("crate://")
refresh_after_dml(engine)
```

```python
import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker
from sqlalchemy_cratedb.support import refresh_after_dml

engine = sa.create_engine("crate://")
session = sessionmaker(bind=engine)()
refresh_after_dml(session)
```

:::{warning}
Refreshing the table after each DML operation can cause serious
performance degradation, and should only be used on low-volume, low-traffic
data, when applicable, and if you know what you are doing.
:::

## Support for pandas and Dask

> Todo.

Reviewer: Maybe remove this if you're not planning to fill in this PR.

Author: I've just updated the patch, with this section completed.

-- https://sqlalchemy-cratedb--28.org.readthedocs.build/support.html#support-for-pandas-and-dask

(support-features)=
(support-utilities)=
# Support Features

The package bundles a few support and utility functions that try to fill a few
gaps you will observe when working with CrateDB, a distributed OLAP database,
since it lacks certain features, usually found in traditional OLTP databases.

A few of the features outlined below are referred to as [polyfills], and
emulate a few functionalities, for example, to satisfy compatibility issues on
downstream frameworks or test suites. You can use them at your disposal, but

> you should know what you are doing, as some of them can seriously impact the

Reviewer 2: ```suggestion
you should know what you are doing, as some of them can seriously impact 
```

Removed "the" from "the performance".

performance.

Other features include efficiency support utilities for 3rd-party frameworks,
which can be used to increase performance, mostly on INSERT operations.

(support-insert-bulk)=
## Bulk Support for pandas and Dask

:::{rubric} Background
:::
CrateDB's [](inv:crate-reference#http-bulk-ops) interface enables efficient

> INSERT, UPDATE, and DELETE operations for batches of data. It allows to issue

Reviewer 2: ```suggestion
INSERT, UPDATE, and DELETE operations for batches of data. It enables
```

Changed to read "It enables bulk operations".

Author: Thank you!

:::{rubric} Utility
:::
- The `refresh_after_dml` utility will configure an SQLAlchemy engine or session
  to automatically invoke `REFRESH TABLE` statements after each DML
  operation (INSERT, UPDATE, DELETE).
- Only relevant (dirty) entities / tables will be considered to be refreshed.

:::{rubric} Synopsis
:::
```python
import sqlalchemy as sa
from sqlalchemy_cratedb.support import refresh_after_dml

engine = sa.create_engine("crate://")
refresh_after_dml(engine)
```

```python
import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker
from sqlalchemy_cratedb.support import refresh_after_dml

engine = sa.create_engine("crate://")
session = sessionmaker(bind=engine)()
refresh_after_dml(session)
```

:::{warning}
Refreshing the table after each DML operation can cause serious performance
degradations, and should only be used on low-volume, low-traffic data,
when applicable, and if you know what you are doing.
:::

(support-unique)=
## Synthetic UNIQUE Constraints

:::{rubric} Background
:::
CrateDB does not provide `UNIQUE` constraints in DDL statements. Because of its
distributed nature, supporting such a feature natively would cause expensive

> database cluster operations, deterring many benefits of using database clusters

Reviewer 2: ```suggestion
database cluster operations, negating many benefits of using database clusters
```

Minor phrasing update.

Author: Thanks again.
